Refactor PyNEST Microcircuit example Potjans_2014

jhnnsnk opened the pull request:
This PR revises code and documentation of the PyNEST Microcircuit and adjusts the example to NEST3.
Summary of the main changes:

- NEST3-specific
  - remove loops on Python level (using NodeCollections and `nest.random` for distributions)
  - use `nodes.set(params)` and `nodes.parameter=value`

- use `'V0_type': 'optimized'` as default initial condition
- file `example.py` is renamed to `run_microcircuit.py`, the folder `data` is now `reference_data`
- revise all functions in detail
  - change the order of functions and parameter definitions in the respective files for a better overview
  - get rid of redundant functions and parameter definitions
  - remove unnecessary loops and nested if conditions
  - revise all function names and update many variable names to be more descriptive
  - store only those parameters as class attributes for which this is needed
  - provide explicit arguments to functions in `helpers.py` instead of dictionaries
  - define only one `tau_syn` instead of `tau_syn_ex` and `tau_syn_in` with the same value (only one `PSC_e_over_PSP_e` was computed anyways)
  - scale number of neurons in raster plot based on `N_scaling`
- new structure of `Network class` with distinction between create and connect and use of private methods
  - disentangle derivation of dependent parameters from create and connect calls by introducing function `__derive_parameters()`
  - synapse numbers are now only once cast as int
- detailed time measurement (total time, initialize, create, connect, simulate, evaluate) on each rank
- update all docstrings, comments and printouts (use `.format()` instead of `%`) as well as file headers (as discussed with @sarakonradi) 
- new reference data generated with default values
- autopep8
- refactor README.rst (as discussed with @sarakonradi)
  - update structure and text (details on parameters and usage)
  - add images
  - add acknowledgments, citation, information about this PyNEST Microcircuit and other versions
  
The following points still need discussions or decisions:

1. The amplitude of DC input for stimulation is multiplied with `K_ext`. Shall we keep that?
2. The number of thalamic neurons is not scaled with `N_scaling`. Is this intended?
3. What is the best practice for naming parameters (e.g., should mean be prefix or suffix)?
4. How were the original values of `full_mean_rates` computed? Shall we replace them by values obtained with this version? If yes, which combination of MPI processes and threads shall we choose?

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
 
+The code can be `parallelized <https://nest-simulator.readthedocs.io/en/latest/guides/parallel_computing.html>`_ using OpenMP and MPI, if NEST has been built with these applications.
+The number of threads (per MPI process) can be chosen by adjusting ``local_num_threads`` in ``sim_params.py``.
+The command for running the script with two MPI processes would be:
```
`with two MPI processes is:`

albada replied:
maybe 'features' instead of 'applications'

jhnnsnk replied:
Done.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+Note on Parameters
+##################
+
+By default, the simulation uses external Poissonian input to excite all neuronal populations of the microcircuit, see ``poisson_input': True`` in ``network_params.py``.
```
Do you mean `microcircuit (see ``poisson_input': True`` in ``network_params.py``)`?

jhnnsnk replied:
Done.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+##################
+
+By default, the simulation uses external Poissonian input to excite all neuronal populations of the microcircuit, see ``poisson_input': True`` in ``network_params.py``.
+If set to ``False``, the Poissonian input is turned off and compensated approximately by calculated DC input.
```
Write `direct current (DC)` the first time `DC` is mentioned

jhnnsnk replied:
Done.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+In addition to this ongoing external drive, a thalamic stimulation or a stimulation by an external DC input can be switched on in ``stimulus_params.py`` (the default for both types of stimuli is ``False``).
+
+The default random initialization of membrane voltages in this simulation uses population-specific means and standard deviations to reduce an initial activity burst in the network: ``'V_type': 'optimized'`` in ``network_params.py``.
+Previous implementations used the same mean and standard deviation for all populations which corresponds to setting ``'V_type': 'original'``.
```
A bit confusing. Do you mean ` Previous implementations used the same mean and standard deviation for all populations by setting ``'V_type': 'original'`` `?

albada replied:
add a comma: populations, which
I don't find this confusing, but maybe 'which is achieved by setting' would be even clearer?

jhnnsnk replied:
I changed it to ``, which is here achieved by setting``. In some previous implementations like the SLI version, the parameter ``V_type`` did not exist at all. In the previous PyNEST version the parameter existed with ``original```as default.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+
+2020: revision of code and documentation, and adaptation to NEST3 by Johanna Senk
+
+Citation
```
I would suggest moving this block somewhere more visible. Perhaps as an admonition at the very top? 

.. admonition:: Citing this code

    If you use this code, we ask you to cite the paper by Potjans and Diesmann [1]_ and the NEST release on Zenodo.

jhnnsnk replied:
I moved it up to the end of the first section below the figures.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+Acknowledgments
+###############
+
+This project has received funding from the European Union Seventh Framework Programme ([FP7/2007-2013]) under grant agreement n° 604102 (Human Brain Project, HBP), and the European Union’s Horizon 2020 Framework Programme for Research and Innovation under Specific Grant Agreement No. 720270 (Human Brain Project SGA1) and No. 785907 (Human Brain Project SGA2).
```
Remove comma: `(Human Brain Project, HBP) and the European Union’s Horizon 2020 Framework Programme`

jhnnsnk replied:
Done.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+Acknowledgments
+###############
+
+This project has received funding from the European Union Seventh Framework Programme ([FP7/2007-2013]) under grant agreement n° 604102 (Human Brain Project, HBP), and the European Union’s Horizon 2020 Framework Programme for Research and Innovation under Specific Grant Agreement No. 720270 (Human Brain Project SGA1) and No. 785907 (Human Brain Project SGA2).
```
Add `Funding for X:` at the start of this paragraph too, for clarity and consistency

jhnnsnk replied:
I said ``Funding for the PyNEST microcircuit:``. Are you fine with this?

sarakonradi replied:
Yes, that's a lot clearer.

sarakonradi commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+Other Implementations of the Microcircuit Model
+###############################################
+
+The original `SLI version <https://github.com/nest/nest-simulator/tree/master/examples/nest/Potjans_2014>`__ by Tobias Potjans and Markus Diesmann (later revised by David Dahmen and Sacha van Albada) is also part of the NEST code base as an example.
```
`The original SLI version by Tobias Potjans and Markus Diesmann ([INSERT YEAR]) is included in the NEST code base as an example. This version was revised by David Dahmen and Sacha van Albada in [INSERT YEAR]`.

albada replied:
The version by Tobias was never included in NEST. Also, Tom helped create the SLI version that is part of NEST. Maybe reword to 'A SLI version by David Dahmen, Tom Tetzlaff, and Sacha van Albada, based on the original version by Tobias Potjans and Markus Diesmann, is also part of the NEST code base as an example.'

jhnnsnk replied:
I used the version by @albada without the years as they are not precisely known.
@albada: Can we also name authors of the PyNN version?

stinebuu commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
-                    }
-                }
+                    'sigma': self.stim_dict['delay_th_std'][i],
+                    'low': self.sim_resolution}}
```
And this should be
```
'delay': nest.math.redraw(nest.random.normal(mean=self.stim_dict['delay_th'][i], std=self.stim_dict['delay_th_std'][i]), low=self.sim_resolution, high=np.Inf)
```

jhnnsnk replied:
Done.

stinebuu commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -168,91 +83,90 @@ def get_std_PSP_matrix(PSP_rel, number_of_pop):
+    # parameters of the neuron model
     'neuron_params': {
-        # Membrane potential average for the neurons (in mV).
+        # membrane qpotential average for the neurons (in mV)
```
```suggestion
        # membrane potential average for the neurons (in mV)
```

jhnnsnk replied:
Fixed.

albada commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
------------
+This is a PyNEST implementation of the microcircuit model by Potjans and Diesmann [1]_.
+The network model represents four layers of cortex, L2/3, L4, L5, and L6, each consisting of two populations of excitatory and inhibitory neurons.
 
```
I would write 'each consisting of a population of excitatory neurons and a population of inhibitory neurons'

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
-      Contains the parameters for the stimuli.
-   -  ``example.py``
-      Use this script to try out the microcircuit.
+.. |img3| image:: reference_data/box_plot.png
```
The firing rate should be in spikes/s, not Hz.

jhnnsnk replied:
I changed it in the source code and it will appear in the plot when I upload a new version.

albada commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+
+2019: optimized initialization of membrane voltages added by Han-Jia Jiang
+
+2020: revision of code and documentation, and adaptation to NEST3 by Johanna Senk
```
Since the current communicating author is at the top, I would order the changes from the bottom up as well, with the most recent ones at the top.

jhnnsnk replied:
Both changes are applied.

albada commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+Acknowledgments
+###############
+
+This project has received funding from the European Union Seventh Framework Programme ([FP7/2007-2013]) under grant agreement n° 604102 (Human Brain Project, HBP), and the European Union’s Horizon 2020 Framework Programme for Research and Innovation under Specific Grant Agreement No. 720270 (Human Brain Project SGA1) and No. 785907 (Human Brain Project SGA2).
```
has received -> received

jhnnsnk replied:
``has received`` is in the official formulation for citing the HBP. @albada, are you sure that we should change it here?

albada commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+
+This project has received funding from the European Union Seventh Framework Programme ([FP7/2007-2013]) under grant agreement n° 604102 (Human Brain Project, HBP), and the European Union’s Horizon 2020 Framework Programme for Research and Innovation under Specific Grant Agreement No. 720270 (Human Brain Project SGA1) and No. 785907 (Human Brain Project SGA2).
+
+Funding for [1]_: This work was supported by the Helmholtz Alliance on Systems Biology; European Union (FACETS, grant 15879 and BrainScaleS, grant 269921); Deutsch-Israelische Projektkooperation (DIP, grant F1.2); Bundesministerium für Bildung und Forschung, Germany (BMBF, grant 01GQ0420 to BCCN Freiburg), and the Next-Generation Supercomputer Project of the Ministry of education, culture, sports ...
```
add comma after Juelich

jhnnsnk replied:
The previous version was copied from the paper but I see that it is better with the comma. It is added now.

albada commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -19,146 +19,61 @@
+    # 'populations')
+    'full_num_neurons':
+        np.array([20683, 5834, 21915, 5479, 4850, 1065, 14395, 2948]),
+    # mean rates of the different populations in the non-scaled version
```
indicate the unit: spikes/s

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -168,91 +83,90 @@ def get_std_PSP_matrix(PSP_rel, number_of_pop):
-    # Number of external connections to the different populations.
-    # The order corresponds to the order in 'populations'.
+             [0.0364, 0.001, 0.0034, 0.0005, 0.0277, 0.008, 0.0658, 0.1443]]),
+    # number of external connections to the different populations (same order
```
number -> indegree

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -168,91 +83,90 @@ def get_std_PSP_matrix(PSP_rel, number_of_pop):
+    # relative inhibitory synaptic strength
     'g': -4,
-    # Rate of the Poissonian spike generator (in Hz).
+    # rate of the Poissonian spike generator (in Hz)
```
Hz -> spikes/s

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -168,91 +83,90 @@ def get_std_PSP_matrix(PSP_rel, number_of_pop):
+    # if False: DC input is applied for compensation
     'poisson_input': True,
-    # Delay of the Poisson generator (in ms).
+    # delay of the Poisson generator (in ms)
```
maybe write 'delay from the Poisson generator to the network (in ms)' for clarity

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

jhnnsnk commented:
I just added a presimulation time with the parameter `t_presim`. This presimulation runs before the main simulation and it can be used to separate the setup of the presynaptic connectivity structure and intialization artifacts from the main simulation for time measurements. The default value is `t_presim = 0.`, meaning that the measured time corresponds only to the setup of the presynaptic connectivity. Please have a look and let me know if you find this confusing, in particular @albada.

albada commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -168,91 +83,90 @@ def get_std_PSP_matrix(PSP_rel, number_of_pop):
+    # mean delay of excitatory connections (in ms)
     'mean_delay_exc': 1.5,
-    # Mean delay of inhibitory connections (in ms).
+    # mean delay of inhibitory connections (in ms)
```
I'm okay with the naming mean_delay_exc, mean_delay_inh, V0_mean, but if you want to consistently have 'mean' either in front or at the back, I would choose at the back: delay_exc_mean etc.

jhnnsnk replied:
I renamed and used the back as you suggested.

albada commented on pynest/examples/Potjans_2014/network_params.py:
```diff
@@ -168,91 +83,90 @@ def get_std_PSP_matrix(PSP_rel, number_of_pop):
+    net_dict['PSP_e'],
+    net_dict['PSP_e'] * net_dict['g'],
+    len(net_dict['populations']))
+mean_PSP_matrix[0, 2] *= 2
```
If someone accidentally called this function twice, the L4E->L23E strength would be quadrupled. Normally this shouldn't happen, but by defining a variable PSP_23e_4e = 2 * PSP_e and then setting mean_PSP_matrix[0, 2] = PSP_23e_4e you avoid this altogether.

jhnnsnk replied:
I did not define a variable, but directly set the matrix entry now to the correct value.
```suggestion
mean_PSP_matrix[0, 2] = 2. * net_dict['PSP_e']
```
Are you fine with this?

jhnnsnk replied:
I changed to `PSP_matrix_mean[0, 2] = 2. * net_dict['PSP_exc_mean']`. Parameter names have also changed. Please let me know if this is fine with you.

albada commented on pynest/examples/Potjans_2014/run_microcircuit.py:
```diff
@@ -0,0 +1,83 @@
+
+###############################################################################
+# Summarize time measurements. Rank 0 usually takes longest because of the
+# data evaluation and printouts.
```
maybe 'print calls' instead of 'printouts' (which normally means printed paper)

jhnnsnk replied:
Done.

jhnnsnk replied:
Changed.

albada commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -19,33 +19,35 @@
+    # simulation time (in ms)
     't_sim': 1000.0,
-    # Resolution of the simulation (in ms).
+    # Resolution of the simulation (in ms)
```
R -> r

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -19,33 +19,35 @@
-    # Path to save the output data.
+    # list of recording devices, default is 'spike_detector'. a 'voltmeter' can
+    # be added to record membrane voltages of the neurons. nothing will be
+    # recorded if an empty list is given.
```
Start full sentences with capital letters.

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -19,49 +19,50 @@
+    # duration of the thalamic input (in ms)
     'th_duration': 10.0,
-    # Rate of the thalamic input (in Hz).
+    # rate of the thalamic input (in Hz)
```
Hz -> spikes/s

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -19,49 +19,50 @@
-    'dc_amp': np.ones(len(net_dict['populations'])) * 0.3,
-    }
+    # mean amplitude of the thalamic postsynaptic potential (in mV),
+    # standard deviation will be taken from 'net_dict'
```
standard deviation -> relative standard deviation

jhnnsnk replied:
Actually it was the absolute standard deviation before. But I changed it now to the relative one and only use one value instead of an array.

jhnnsnk replied:
In fact, it was the absolute standard deviation before, but changed it now for consistency. Please also check how I now got rid of the matrices for relative standard deviations and use scalars instead.

albada commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -19,49 +19,50 @@
+    # optional DC input
+    # turn DC input on or off (True or False)
+    'dc_input': False,
+    # start of the DC generator (in ms)
```
DC generator -> DC input

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -19,49 +19,50 @@
+    'dc_input': False,
+    # start of the DC generator (in ms)
+    'dc_start': 0.0,
+    # duration of the DC generator (in ms)
```
DC generator -> DC input

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

albada commented on pynest/examples/Potjans_2014/stimulus_params.py:
```diff
@@ -19,49 +19,50 @@
+    'dc_start': 0.0,
+    # duration of the DC generator (in ms)
+    'dc_dur': 1000.0,
+    # amplitude of the DC generator (in pA),
```
DC generator -> DC input

jhnnsnk replied:
Done.

jhnnsnk replied:
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+######################
 
+By default, the variables ``N_scaling`` and ``K_scaling`` in ``network_params.py`` are set to
+`0.1` which is a good choice for running the microcircuit on a local machine.
```
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
-Microcircuit Example
-=====================
+PyNEST Microcircuit
```
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
-Hendrik Rothe, Hannah Bos, Sacha van Albada
+   python run_microcircuit.py
 
+The output will be saved in the directory ``data``.
```
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/README.rst:
```diff
@@ -1,61 +1,108 @@
+Acknowledgments
+###############
+
+This project has received funding from the European Union Seventh Framework Programme ([FP7/2007-2013]) under grant agreement n° 604102 (Human Brain Project, HBP), and the European Union’s Horizon 2020 Framework Programme for Research and Innovation under Specific Grant Agreement No. 720270 (Human Brain Project SGA1) and No. 785907 (Human Brain Project SGA2).
```
I do not know of any global decision. For the microcircuit, @diesmann, @albada and me thought it would be a good idea to be in line with the [multi-area model](https://github.com/INM-6/multi-area-model/blob/master/README.md).

jhnnsnk replied:
I do not know of any global decision. We (@diesmann, @albada and me) just thought it would be ok here to be in line with the [multi-area-model](https://github.com/INM-6/multi-area-model/blob/master/README.md).

jhnnsnk commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -19,315 +19,176 @@
-
-def get_total_number_of_synapses(net_dict):
-    """ Returns the total number of synapses between all populations.
+    PSC_over_PSP = (((C_m) ** (-1) * tau_m * tau_syn / (
```
As far as I know, this equation is not mentioned in any paper, yet. I revised the code to make it more readable and added an explanation on how the equation is obtained in the docstring.

diesmann replied:
The expression for tmax appears in the appendix of Hanuschkin et al. (2010). From this V(tmax) can be easily computed. The Hanuschkin et al. (2010) reference is the best we can do for now. Also Krishnan et al. (2018) cites Hanuschkin in this way.

Hanuschkin A, Kunkel S, Helias M, Morrison A and Diesmann M (2010) A general and efficient method for incorporating precise spike times in globally time-driven simulations. Front. Neuroinform. 4:113. doi: 10.3389/fninf.2010.00113

Krishnan J, Porta Mana P, Helias M, Diesmann M and Di Napoli E (2018) Perfect Detection of Spikes in the Linear Sub-threshold Dynamics of Point Neurons. Front. Neuroinform. 11:75. doi: 10.3389/fninf.2017.00075

jhnnsnk replied:
Thanks, @diesmann! I added the reference. In addition, I changed the function so that it returns the conversion factor `PSC_over_PSP` and has to be called only once. Please check.

jhnnsnk commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -336,50 +197,50 @@ def plot_raster(path, name, begin, end):
         Name of the spike detector.
     begin
-        Initial value of spike times to plot.
+        Time point (in ms) to start plotting spikes (included).
```
Yes. Do you have a suggestion how to make this clearer?

stinebuu replied:
No, so lets leave it as is :)

jhnnsnk commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -19,315 +19,176 @@
+        External weight (in pA).
+    tau_syn
+        Synaptic time constant (in ms).
+    full_mean rates
```
Fixed.

jhnnsnk commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -19,315 +19,176 @@
+        External weight (in pA).
+    tau_syn
+        Synaptic time constant (in ms).
+    full_mean rates
```
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -19,315 +19,176 @@
-        Lowest and highest ids of the spike detectors.
+    # recurrent input of full network
+    indegree_matrix = \
+        full_num_synapses / full_num_neurons.reshape(len(K_ext), 1)
```
Yes, it is doing exactly the same. Thanks.

jhnnsnk commented on pynest/examples/Potjans_2014/helpers.py:
```diff
@@ -388,81 +249,169 @@ def fire_rate(path, name, begin, end):
+    for l in node_idfile:
+        a = l.split()
+        node_ids.append([int(a[0]), int(a[1])])
+    return sd_files, sd_names, np.array(node_ids)
```
I set the dtype and got rid of the `int()` so that it looks less cumbersome.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
         if stim_dict is not None:
             self.stim_dict = stim_dict
         else:
             self.stim_dict = None
```
Obviously, thanks. That was still a remainder from the old code.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
-            )
+            os.path.join(self.data_path, 'population_nodeids.dat'), 'w+')
+
         for i, pop in enumerate(self.net_dict['populations']):
```
True. For that purpose I actually introduced `self.num_pops`. I changed to:
```suggestion
        for i in np.arange(self.num_pops):
```

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
-            pop_file.write('%d  %d \n' % (
-                population[0].get('global_id'),
-                population[-1].get('global_id')))
+            pop_file.write('{} {}\n'.format(population.global_id[0],
```
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+        Thalamic neurons are of type ``parrot_neuron`` and receive input from a
+        Poisson generator.
+        Note that the number of thalamic neurons is not scaled with
+        ``Ǹ_scaling``.
```
How to pronounce that even? Fixed.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+                            'distribution': 'normal_clipped',
+                            'mu': self.mean_weight_matrix[i][j],
+                            'sigma': abs(self.mean_weight_matrix[i][j] *
+                                         self.std_weight_matrix[i][j])},
```
I changed to `redraw`, but I had to use the keywords `min` and `max`. It works with `np.Inf` and `np.NINF`.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
-                    if weight < 0:
+                            'mu': self.net_dict['mean_delay_matrix'][i][j],
+                            'sigma': self.net_dict['std_delay_matrix'][i][j],
+                            'low': self.sim_resolution}}
```
Done (but with `min` and `max` here as well).

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+                            'sigma': self.net_dict['std_delay_matrix'][i][j],
+                            'low': self.sim_resolution}}
+
+                    if self.mean_weight_matrix[i][j] < 0:
```
Done. This part is moved above.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
-                    },
+                    'mu': self.weight_th,
+                    'sigma': self.weight_th * self.net_dict['PSP_std'],
+                    'low': 0.0},
```
Done.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+
+        # scaled numbers of neurons and synapses
+        self.num_neurons = (self.net_dict['full_num_neurons'] *
+                            self.net_dict['N_scaling']).astype(int)
```
What you suggest is the better choice. I modified accordingly. This changes simulation results. Before merging, I will upload the latest data.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+                self.net_dict['bg_rate'], self.net_dict['K_ext'],
+                self.net_dict['neuron_params']['tau_syn'],
+                PSC_ext)
+            print(DC_amp)
```
This was a leftover from testing. I don't think that DC_amp needs to be printed, so I removed the statement.

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+            print('Creating neuronal populations.')
+
         self.pops = []
         pop_file = open(
```
Now I write all node ids to file on Rank 0 after all populations are created. The benefit is that only Rank 0 has to do it, but another loop is needed. Before, I did not want to use `with open()` because I did not want to indent the code block any further, but this solution seems to be better. What do you think?

jhnnsnk commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,502 @@
+                population.set(V_m=nest.random.normal(
+                    self.net_dict['neuron_params']['V0_mean']['original'],
+                    self.net_dict['neuron_params']['V0_std']['original']))
+
```
Done.

AlexVanMeegen commented on pynest/examples/Potjans_2014/network.py:
```diff
@@ -19,486 +19,522 @@
+            if self.net_dict['K_scaling'] != 1:
+                num_th_synapses *= self.net_dict['K_scaling']
+                self.weight_th /= np.sqrt(self.net_dict['K_scaling'])
+            self.num_th_synapses = num_th_synapses.astype(int)
```
Here is still a 'floor-cast-to-int' instead of the `np.round(x).astype(int)`

jhnnsnk replied:
Thanks. It is fixed now.

AlexVanMeegen commented on pynest/examples/Potjans_2014/sim_params.py:
```diff
@@ -19,33 +19,35 @@
-    'master_seed': 55,
-    # Number of threads per MPI process.
+    # masterseed for NEST and NumPy
+    'master_seed': 555,
```
Bump ;)

jhnnsnk replied:
I changed it for testing, but now I changed it back.

sarakonradi commented:
@steffengraber and I were discussing whether or not it makes sense to include more information on the microcircuit model itself. Currently, there is one sentence. 

@jhnnsnk What do you think?

jhnnsnk commented:
Here are my responses to @albada's comments:
>     1. Multiplying the transient DC amplitude with K_ext seems a bit more complicated than necessary to me. I think it would be more straightforward to make it equal across populations.
I just did some testing: If we use just one DC amplitude, the relative fraction of firing rates in the microcircuit gets destroyed. If we multiply with `K_ext`, we can increase all rates more or less equally. The latter would be treated as the ongoing DC input (for compensation of Poisson input and adjusted if the network is scaled), with the difference that starting point, duration, and amplitude can be set manually. The more I think about it, the more I prefer to keep the population-specific DC input. I changed starting point and duration so that the DC input shows up in the middle interval of the raster plot. Please let me know if this is ok.

>     2. The lack of scaling of the thalamic population is on purpose, because it's a small population and this is intended to avoid unnecessary distortions of the dynamics.
Ok.
 
>     3. I left a comment about whether to use 'mean' as prefix or suffix. I think that the latter is slightly better, because it makes the beginnings of parameter names more unique, which helps with automatic name completion in interactive mode.
I changed accordingly, not only for weights but also for delays.

>     4. I don't remember exactly how full_mean_rates were computed, but I think it was with PyNN. Just to be able to write down exactly how they were computed, and for consistency, I think it is a good idea to compute them with the current version of the code, of course discarding enough of a transient.
Good. I just did a run of the full network and adjusted the values. I documented how the result was obtained.

jhnnsnk commented:
> @steffengraber and I were discussing whether or not it makes sense to include more information on the microcircuit model itself. Currently, there is one sentence.
> 
> @jhnnsnk What do you think?

I do not think that much more general information is needed. What do you think, @albada?

albada commented:
> > @steffengraber and I were discussing whether or not it makes sense to include more information on the microcircuit model itself. Currently, there is one sentence.
> > @jhnnsnk What do you think?
> 
> I do not think that much more general information is needed. What do you think, @albada?

@jhnnsnk I agree with you. People can refer to the paper to find out more about the model.

jhnnsnk commented:
I added "Recommendations for benchmarking" to the README as discussed with @diesmann.

jhnnsnk commented:
I added a list of collaborators to the README in the contribution section. @sarakonradi @diesmann @AlexVanMeegen @stinebuu @hakonsbm @albada: Please check!

albada commented:
I'm happy with the changes to the README.

stinebuu commented:
I am also happy with the changes

H27CK commented:
Quick note: my experience with this refactor is positive. In the configuration suggested for benchmarking, all operates as expected on ~10 nodes of JURECA. Initial tests show very good performance, relative to that observed when using the case/code combination distributed with 2.18.1.
